modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -23,48 +37,132 @@ import eu.timepit.refined.types.numeric.{NonNegLong, PosLong}
   ): GlobalL0Service[F] =
     new GlobalL0Service[F] {
 
+      private val numConcurrentQueries = 10
       private val logger = Slf4jLogger.getLogger[F]
+      private val maybeMajorityPeerIds = maybeMajorityPeerIdSet.map(_.toNonEmptyList)
 
-      def pullLatestSnapshot: F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] =
-        globalL0ClusterStorage.getRandomPeer.flatMap { l0Peer =>
-          l0GlobalSnapshotClient.getLatest(l0Peer).flatMap {
-            case ((snapshot, state)) =>
-              snapshot.toHashedWithSignatureCheck.flatMap(_.liftTo[F]).map((_, state))
-          }
-        }
+      def pullLatestSnapshot: F[LatestSnapshotTuple] =
+        maybeMajorityPeerIds.fold(pullLatestSnapshotFromRandomPeer)(pullLatestSnapshotWithMajorityHash)
 
       def pullGlobalSnapshot(hash: Hash): F[Option[Hashed[GlobalIncrementalSnapshot]]] =
         pullGlobalSnapshot(l0GlobalSnapshotClient.get(hash)).handleErrorWith { e =>
           logger
-            .warn(e)(s"Failure pulling single snapshot with hash=$hash")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .warn(e)(s"Failure pulling single snapshot with hash=${hash.show}")
+            .as(none)
         }
 
       def pullGlobalSnapshot(ordinal: SnapshotOrdinal): F[Option[Hashed[GlobalIncrementalSnapshot]]] =
         pullGlobalSnapshot(l0GlobalSnapshotClient.get(ordinal)).handleErrorWith { e =>
           logger
-            .warn(e)(s"Failure pulling single snapshot with ordinal=$ordinal")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .warn(e)(s"Failure pulling single snapshot with ordinal=${ordinal.show}")
+            .as(none)
+        }
+
+      def pullGlobalSnapshots: F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        maybeMajorityPeerIds
+          .fold(pullGlobalSnapshotsFromRandomPeer)(pullGlobalSnapshotsFromMajority)
+          .handleErrorWith { e =>
+            logger.warn(e)(s"Failure pulling global snapshots") >>
+              Applicative[F].pure(
+                List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
+              )
+          }
+
+      private def pullGlobalSnapshotsFromMajority(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        lastGlobalSnapshotStorage.getOrdinal.flatMap {
+          _.fold {
+            pullLatestSnapshotWithMajorityHash(majorityPeerIds).map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+          } { lastStoredOrdinal =>
+            for {
+              snapshotData <- getLatestMajoritySnapshotData(majorityPeerIds)
+              l0Peers <- globalL0ClusterStorage.getPeers
+              lastOrdinal = snapshotData.ordinal
+              majorityPeer <- getPeerInTheMajority(lastOrdinal, snapshotData.hash, l0Peers)
+              pulled <- pullSnapshots(majorityPeer, lastStoredOrdinal.next, lastOrdinal)
```

Comment:
The verification if the downloaded data actually matches the majority calculation is still missing. A node can declare/return a hash at ordinal matching the majority one, but it's not guaranteed that the data that node sends is actually correct. You need to verify the snapshots received here - zero trust policy.